Source Versioning docs for MySQL (#36164)
MaterializeInc/materialize: doc/user/content/sql/create-source/mysql-v2.md

pagerank: 40
menu:
  main:
    parent: 'create-source'
    identifier: cs_mysql-v2
    name: MySQL (New Syntax)
    weight: 21
---

{{< private-preview />}}

{{< source-versioning-disambiguation is_new=true
other_ref="[old reference page](/sql/create-source/mysql/)" include_blurb=true >}}

## Prerequisites

{{% create-source/intro %}}
Materialize supports MySQL (5.7+) as a real-time data source. To connect to a
MySQL database, you first need to tweak its configuration to enable
[GTID-based binary log (binlog) replication](#change-data-capture) and set
[`binlog_row_metadata=FULL`](#change-data-capture), and then
[create a connection](#prerequisite-creating-a-connection-to-mysql) in
Materialize that specifies access and authentication parameters.
{{% /create-source/intro %}}

## Syntax

{{% include-syntax file="examples/create_source_mysql" example="syntax" %}}

## Ingesting data

After a source is created, you can create tables from the source referencing
upstream MySQL tables that have GTID-based binlog replication enabled. You can
create multiple tables that reference the same upstream table.

See [`CREATE TABLE FROM SOURCE`](/sql/create-table/) for details.

> #### Handling table schema changes

Reviewer: [rule_020](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_020.yaml) (section-split, conf 2): heading level skip — `## Ingesting data` jumps to `####` with no intervening `###`. Promoting to `###` makes this a peer of `### Change data capture` and `### Monitoring source progress`.

```suggestion
### Handling table schema changes
```

The use of `CREATE SOURCE` with the new [`CREATE TABLE FROM
SOURCE`](/sql/create-table/) allows for the handling of certain upstream DDL
changes without downtime.

See [`CREATE TABLE FROM SOURCE`](/sql/create-table/#handling-table-schema-changes) for details.

> #### Supported types

Reviewer: [rule_020](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_020.yaml) (section-split, conf 2): heading level skip — promote to `###` for consistent hierarchy.

```suggestion
### Supported types
```

With the new syntax, after a MySQL source is created, you [`CREATE TABLE FROM
SOURCE`](/sql/create-table/) to create a corresponding table in Materialize and
start ingesting data.

{{% include-from-yaml data="mysql_source_details"
name="mysql-supported-types" %}}

{{% include-from-yaml data="mysql_source_details"
name="mysql-unsupported-types" %}}

For more information, including strategies for handling unsupported types,
see [`CREATE TABLE FROM SOURCE`](/sql/create-table/).

> #### Upstream table truncation restrictions

Reviewer: [rule_020](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_020.yaml) (section-split, conf 2): heading level skip — promote to `###` for consistent hierarchy.

```suggestion
### Upstream table truncation restrictions
```

{{% include-from-yaml data="mysql_source_details"
name="mysql-truncation-restriction" %}}

For additional considerations, see also [`CREATE TABLE`](/sql/create-table/).

### Change data capture

> {{< note >}}

Reviewer: [rule_043](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_043.yaml) (admonition-rewrap, conf 5): blank line missing after opening tag.

```suggestion
{{< note >}}

```

For step-by-step instructions on enabling GTID-based binlog replication for your
MySQL service, see the integration guides:
[Amazon RDS](/ingest-data/mysql/amazon-rds/),
[Amazon Aurora](/ingest-data/mysql/amazon-aurora/),
[Azure DB](/ingest-data/mysql/azure-db/),
[Google Cloud SQL](/ingest-data/mysql/google-cloud-sql/),
[Self-hosted](/ingest-data/mysql/self-hosted/).

> {{< /note >}}

Reviewer: [rule_043](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_043.yaml) (admonition-rewrap, conf 5): blank line missing before closing tag.

```suggestion

{{< /note >}}
```

The source uses MySQL's binlog replication protocol to **continually ingest
changes** resulting from `INSERT`, `UPDATE` and `DELETE` operations in the
upstream database. This process is known as _change data capture_.

The replication method used is based on [global transaction identifiers
(GTIDs)](https://dev.mysql.com/doc/refman/8.0/en/replication-gtids.html), and
guarantees **transactional consistency** — any operation inside a MySQL
transaction is assigned the same timestamp in Materialize, which means that the
source will never show partial results based on partially replicated
transactions.

Before creating a source in Materialize, you **must** configure the upstream
MySQL database for GTID-based binlog replication:

{{% mysql-direct/ingesting-data/mysql-configs %}}

In addition, the new `CREATE TABLE FROM SOURCE` syntax requires full row
metadata in the binlog. You must set the following system variable on the
upstream MySQL server:

```sql
SET GLOBAL binlog_row_metadata = FULL;
```

If you're running MySQL using a managed service, additional configuration
changes might be required. For step-by-step instructions on enabling GTID-based

> binlog replication for your MySQL service, see the integration guides.

Reviewer: [rule_013](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_013.yaml) (tone-tightening, conf 2): shorten verbose cross-reference.

```suggestion
If you're running MySQL using a managed service, additional configuration
changes might be required. To enable GTID-based binlog replication for your
MySQL service, see the integration guides.
```

#### Binlog retention

> {{< warning >}}

Reviewer: [rule_043](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_043.yaml) (admonition-rewrap, conf 5): blank line missing after opening tag.

```suggestion
{{< warning >}}

```

If Materialize tries to resume replication and finds GTID gaps due to missing
binlog files, the source enters an errored state and you have to drop and
recreate it.

> {{< /warning >}}

Reviewer: rule_043 (admonition-rewrap, conf 5): blank line missing before closing tag.

```suggestion

{{< /warning >}}
```

> {{< /warning >}}

Reviewer: [rule_043](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_043.yaml) (admonition-rewrap, conf 5): blank line missing before closing tag.

```suggestion

{{< /warning >}}
```

[`binlog_expire_logs_seconds`](https://dev.mysql.com/doc/mysql-replication-excerpt/8.0/en/replication-options-binary-log.html#sysvar_binlog_expire_logs_seconds)
system variable. We recommend using the default value for this configuration in
order to not compromise Materialize's ability to resume replication in case of
failures or restarts.

In some MySQL managed services, binlog expiration can be overridden by a
service-specific configuration parameter. It's important that you double-check
if such a configuration exists, and ensure it's set to the maximum interval
available.

As an example, [Amazon RDS for MySQL](/ingest-data/mysql/amazon-rds/) has its
own configuration parameter for binlog retention ([`binlog retention hours`](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/mysql-stored-proc-configuring.html#mysql_rds_set_configuration-usage-notes.binlog-retention-hours))
that overrides `binlog_expire_logs_seconds` and is set to `NULL` by default.

### Monitoring source progress

[//]: # "TODO(morsapaes) Replace this section with guidance using the new
progress metrics in mz_source_statistics + console monitoring, when available

> (also for PostgreSQL)."

Reviewer: [rule_032](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_032.yaml) (comment-removal, conf 5): `TODO(morsapaes)` is a contributor name, not a tracked ticket — the carve-out for tracked issues does not apply.

```suggestion
```

-------------------|---------------------------------------------------------|--------------
`source_id_lower`  | [`uuid`](/sql/types/uuid/)                              | The lower-bound GTID `source_id` of the GTIDs covered by this range.
`source_id_upper`  | [`uuid`](/sql/types/uuid/)                              | The upper-bound GTID `source_id` of the GTIDs covered by this range.
`transaction_id`   | [`uint8`](/sql/types/uint/#uint8-info)                  | The `transaction_id` of the next GTID possible from the GTID `source_id`s covered by this range.

And can be queried using:

```mzsql
SELECT transaction_id
FROM <src_name>_progress;
```

Progress metadata is represented as a [GTID set](https://dev.mysql.com/doc/refman/8.0/en/replication-gtids-concepts.html)
of future possible GTIDs, which is similar to the
[`gtid_executed`](https://dev.mysql.com/doc/refman/8.0/en/replication-options-gtids.html#sysvar_gtid_executed)
system variable on a MySQL replica. The reported `transaction_id` should
increase as Materialize consumes **new** binlog records from the upstream MySQL
database. For more details on monitoring source ingestion progress and debugging

> related issues, see [Troubleshooting](/ops/troubleshooting/).

Reviewer: [rule_013](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_013.yaml) (tone-tightening, conf 2): shorten verbose cross-reference.

```suggestion
database. For more information, see [Troubleshooting](/ops/troubleshooting/).
```

## Known limitations

{{% include-headless "/headless/mysql-considerations" %}}

## Example

> {{< important >}}

Reviewer: [rule_043](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_043.yaml) (admonition-rewrap, conf 5): blank line missing after opening tag.

```suggestion
{{< important >}}

```

Before creating a MySQL source, you must enable GTID-based binlog replication in
the upstream database. For step-by-step instructions, see the integration guide
for your MySQL service: [Amazon RDS](/ingest-data/mysql/amazon-rds/),
[Amazon Aurora](/ingest-data/mysql/amazon-aurora/),
[Azure DB](/ingest-data/mysql/azure-db/),
[Google Cloud SQL](/ingest-data/mysql/google-cloud-sql/),
[Self-hosted](/ingest-data/mysql/self-hosted/).

> {{< /important >}}

Reviewer: [rule_043](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_043.yaml) (admonition-rewrap, conf 5): blank line missing before closing tag.

```suggestion

{{< /important >}}
```

### Creating a source {#create-source-example}

#### Prerequisite: Creating a connection to MySQL

First, you must create a connection to your MySQL database. A connection
describes how to connect and authenticate to an external system you want
Materialize to read data from.

Once created, a connection is **reusable** across multiple `CREATE SOURCE`
statements. For more details on creating connections, check the

> [`CREATE CONNECTION`](/sql/create-connection/#mysql) documentation page.

Reviewer: [rule_013](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_013.yaml) (tone-tightening, conf 2): shorten verbose cross-reference.

```suggestion
Once created, a connection is **reusable** across multiple `CREATE SOURCE`
statements. For more information, see [`CREATE CONNECTION`](/sql/create-connection/#mysql).
```

```mzsql
CREATE SECRET mysqlpass AS '<MYSQL_PASSWORD>';

CREATE CONNECTION mysql_connection TO MYSQL (
    HOST 'instance.foo000.us-west-1.rds.amazonaws.com',
    PORT 3306,
    USER 'materialize',
    PASSWORD SECRET mysqlpass
);
```

If your MySQL server is not exposed to the public internet, you can [tunnel the
connection](/sql/create-connection/#network-security-connections) through an
AWS PrivateLink service (Materialize Cloud) or an SSH bastion host.

{{< tabs tabID="1" >}}
{{< tab "AWS PrivateLink (Materialize Cloud)">}}

{{< include-md file="shared-content/aws-privatelink-cloud-only-note.md" >}}

```mzsql
CREATE CONNECTION privatelink_svc TO AWS PRIVATELINK (
   SERVICE NAME 'com.amazonaws.vpce.us-east-1.vpce-svc-0e123abc123198abc',
   AVAILABILITY ZONES ('use1-az1', 'use1-az4')
);

CREATE CONNECTION mysql_connection TO MYSQL (
    HOST 'instance.foo000.us-west-1.rds.amazonaws.com',
    PORT 3306,
    USER 'root',
    PASSWORD SECRET mysqlpass,
    AWS PRIVATELINK privatelink_svc
);
```

For step-by-step instructions on creating AWS PrivateLink connections and
configuring an AWS PrivateLink service to accept connections from Materialize,

> check [this guide](/ops/network-security/privatelink/).

Reviewer: [rule_013](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_013.yaml) (tone-tightening, conf 2): shorten verbose cross-reference.

```suggestion
For more information, see [this guide](/ops/network-security/privatelink/).
```

{{< /tab >}}
{{< tab "SSH tunnel">}}

```mzsql
CREATE CONNECTION ssh_connection TO SSH TUNNEL (
    HOST 'bastion-host',
    PORT 22,
    USER 'materialize'
);
```

```mzsql
CREATE CONNECTION mysql_connection TO MYSQL (
    HOST 'instance.foo000.us-west-1.rds.amazonaws.com',
    SSH TUNNEL ssh_connection
);
```

For step-by-step instructions on creating SSH tunnel connections and configuring
an SSH bastion server to accept connections from Materialize, check

> [this guide](/ops/network-security/ssh-tunnel/).

Reviewer: [rule_013](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_013.yaml) (tone-tightening, conf 2): shorten verbose cross-reference.

```suggestion
For more information, see [this guide](/ops/network-security/ssh-tunnel/).
```

{{< /tab >}}
{{< /tabs >}}

#### Creating the source in Materialize

You **must** enable GTID-based binlog replication before creating the source.
See [Change data capture](#change-data-capture) for configuration details.

Once replication is configured, create a `SOURCE` in Materialize:

```mzsql
CREATE SOURCE mz_source
    FROM MYSQL CONNECTION mysql_connection;
```

After a source is created, you can create a table from the source, referencing
specific upstream table(s).

_Creates a table in Materialize from the upstream table `mydb.orders`_

```mzsql
CREATE TABLE orders FROM SOURCE mz_source (REFERENCE mydb.orders);
```

> 

Reviewer: [rule_052](https://github.com/MaterializeInc/mz-skills/blob/main/.claude/skills/mz-doc-edit/rules/rule_052.yaml) (callout-insertion, conf 5): MySQL CDC sources should document `ALTER SOURCE` snapshot-blocking behavior via the shared include. The include file exists at `doc/user/content/headless/alter-source-snapshot-blocking-behavior.md`.

```suggestion

### Modifying an existing source

{{< include-md file="headless/alter-source-snapshot-blocking-behavior.md" >}}

```

    AWS PRIVATELINK privatelink_svc
);
```

For more information, see [this guide](/ops/network-security/privatelink/).

{{< /tab >}}
{{< tab "SSH tunnel">}}

```mzsql
CREATE CONNECTION ssh_connection TO SSH TUNNEL (
    HOST 'bastion-host',
    PORT 22,
    USER 'materialize'
);
```

```mzsql
CREATE CONNECTION mysql_connection TO MYSQL (
    HOST 'instance.foo000.us-west-1.rds.amazonaws.com',
    SSH TUNNEL ssh_connection
);
```

For more information, see [this guide](/ops/network-security/ssh-tunnel/).

{{< /tab >}}
{{< /tabs >}}

#### Creating the source in Materialize

You **must** enable GTID-based binlog replication before creating the source.
See [Change data capture](#change-data-capture) for configuration details.

Once replication is configured, create a `SOURCE` in Materialize:

```mzsql
CREATE SOURCE mz_source
    FROM MYSQL CONNECTION mysql_connection;
```

After a source is created, you can create a table from the source, referencing
specific upstream table(s).

_Creates a table in Materialize from the upstream table `mydb.orders`_

```mzsql
CREATE TABLE orders FROM SOURCE mz_source (REFERENCE mydb.orders);
```

### Modifying an existing source

> {{< include-md file="headless/alter-source-snapshot-blocking-behavior.md" >}}

Reviewer 2: The wrong shortcode

https://preview.materialize.com/materialize/36164/sql/create-source/mysql-v2/#modifying-an-existing-source

<img width="775" height="309" alt="Image" src="https://github.com/user-attachments/assets/4c318587-b06f-4015-9f43-ab92b619ea9a" />

Author: ah whoops, that was an LLM suggestion gone wrong. Just going to delete this section as that is related to `ALTER SOURCE` and is documented in those docs already, and corresponding docs for postgres v2 and sql server v2 don't have this section..

---
title: "CREATE SOURCE: MySQL (New Syntax)"
description: "Connecting Materialize to a MySQL database for Change Data Capture (CDC)."
pagerank: 40
menu:
  main:
    parent: 'create-source'
    identifier: cs_mysql-v2
    name: MySQL (New Syntax)
    weight: 21
---

> {{< private-preview />}}

Reviewer 2: public preview

{{< source-versioning-disambiguation is_new=true
other_ref="[old reference page](/sql/create-source/mysql/)" include_blurb=true >}}

## Prerequisites

{{% create-source/intro %}}

> Materialize supports MySQL (5.7+) as a real-time data source. To connect to a

Reviewer 3: Supported version should be 8.0.1?

> MySQL database, you first need to tweak its configuration to enable

Reviewer 3: Should this go into `doc/user/layouts/shortcodes/mysql-direct/ingesting-data/mysql-configs.html` instead?  I'm not sure how we disambiguate settings for new syntax vs. old, probably need input from @kay-kim

Reviewer 2: You sir are correct in flagging the clean disambiguation between the old and new syntax.

This is a separate page for the new syntax ... but this page seems to be  mixing both new and old.
For example ... this syntax ... includes the old syntax options:

<img width="838" height="1017" alt="Image" src="https://github.com/user-attachments/assets/a58f0ffc-dbfd-4410-b80a-a26100246cf6" />

We also didn't seem to have updated the https://preview.materialize.com/materialize/36164/sql/create-table/ page either.

Reviewer 2: We also didn't seem to have updated these pages either.

- https://preview.materialize.com/materialize/36164/sql/create-source/ page 

- https://preview.materialize.com/materialize/36164/sql/create-table/

[Azure DB](/ingest-data/mysql/azure-db/),
[Google Cloud SQL](/ingest-data/mysql/google-cloud-sql/),
[Self-hosted](/ingest-data/mysql/self-hosted/).

{{< /note >}}

The source uses MySQL's binlog replication protocol to **continually ingest
changes** resulting from `INSERT`, `UPDATE` and `DELETE` operations in the
upstream database. This process is known as _change data capture_.

The replication method used is based on [global transaction identifiers
(GTIDs)](https://dev.mysql.com/doc/refman/8.0/en/replication-gtids.html), and
guarantees **transactional consistency** — any operation inside a MySQL
transaction is assigned the same timestamp in Materialize, which means that the
source will never show partial results based on partially replicated
transactions.

Before creating a source in Materialize, you **must** configure the upstream
MySQL database for GTID-based binlog replication:

{{% mysql-direct/ingesting-data/mysql-configs %}}

In addition, the new `CREATE TABLE FROM SOURCE` syntax requires full row
metadata in the binlog. You must set the following system variable on the
upstream MySQL server:

```sql
SET GLOBAL binlog_row_metadata = FULL;

> ```

Reviewer 3: Does this `SET` persist it, if not, may want to redirect them to the parameter config?

`source_id_lower`  | [`uuid`](/sql/types/uuid/)                              | The lower-bound GTID `source_id` of the GTIDs covered by this range.
`source_id_upper`  | [`uuid`](/sql/types/uuid/)                              | The upper-bound GTID `source_id` of the GTIDs covered by this range.
`transaction_id`   | [`uint8`](/sql/types/uint/#uint8-info)                  | The `transaction_id` of the next GTID possible from the GTID `source_id`s covered by this range.

And can be queried using:

```mzsql
SELECT transaction_id
FROM <src_name>_progress;
```

Progress metadata is represented as a [GTID set](https://dev.mysql.com/doc/refman/8.0/en/replication-gtids-concepts.html)
of future possible GTIDs, which is similar to the
[`gtid_executed`](https://dev.mysql.com/doc/refman/8.0/en/replication-options-gtids.html#sysvar_gtid_executed)
system variable on a MySQL replica. The reported `transaction_id` should
increase as Materialize consumes **new** binlog records from the upstream MySQL
database. For more details on monitoring source ingestion progress and debugging
related issues, see [Troubleshooting](/ops/troubleshooting/).

## Known limitations

> {{% include-headless "/headless/mysql-considerations" %}}

Reviewer 2: Doesn't this include the legacy syntax' schema change limitations?

<img width="750" height="493" alt="Image" src="https://github.com/user-attachments/assets/d2e20251-a3d1-4576-8766-3a8390df33d5" />
